be/src/olap/aggregate_func.h

```diff
@@ -15,221 +15,283 @@
+#pragma once
 
 #include "olap/hll.h"
 #include "olap/types.h"
 #include "util/arena.h"
 
 namespace doris {
 
-using AggregateFunc = void (*)(char* left, const char* right, Arena* arena);
-using FinalizeFunc = void (*)(char* data);
+using AggeInitFunc = void (*)(char* dst, Arena* arena);
+using AggUpdateFunc = void (*)(char* dst, const char* src, Arena* arena);
+using AggFinalizeFunc = void (*)(char* data, Arena* arena);
 
-template<FieldAggregationMethod agg_method,
-        FieldType field_type> struct AggregateFuncTraits {};
+// This class contains information about aggregate operation.
+class AggregateInfo {
+public:
+    // Init function will initialize aggregation execute environment in dst.
+    // For example: for sum, we just initial dst to 0. For HLL column, it will
+    // allocate and init context used to compute HLL.
+    //
+    // Memory Note: For plain memory can be allocated from arena, whose lifetime
+    // will last util finalize function is called. Memory allocated from heap should
+    // be freed in finalize functioin to avoid memory leak.
+    inline void init(void* dst, Arena* arena) const {
+        _init_fn((char*)dst, arena);
+    }
+
+    // Actually do the aggregate operation. dst is the context which is initialized
+    // by init function, src is the current value which is to be aggregated.
+    // For example: For sum, dst is the current sum, and src is the next value which
+    // will be added to sum.
+    // This function usually is used when load function.
+    //
+    // Memory Note: Same with init function.
+    inline void update(void* dst, const void* src, Arena* arena) const {
+        _update_fn((char*)dst, (const char*)src, arena);
+    }
+
+    // Merge aggregated intermediate data. Data stored in engine is aggregated,
+    // because storage has done some aggregate when loading or compaction.
+    // So this function is often used in read operation.
+    // 
+    // Memory Note: Same with init function.
+    inline void merge(void* dst, const void* src, Arena* arena) const {
+        _merge_fn((char*)dst, (const char*)src, arena);
+    }
+
+    // Finalize function convert intermediate context into final format. For example:
+    // For HLL type, finalize function will serialize the aggregate context into a slice.
+    // For input src points to the context, and when function is finished, result will be
+    // saved in src.
+    //
+    // Memory Note: All heap memory allocated in init and update function should be freed
+    // before this function return. Memory allocated from arena will be still available
+    // and will be freed by client.
+    inline void finalize(void* src, Arena* arena) const {
+        _finalize_fn((char*)src, arena);
+    }
 
-template<FieldType field_type>
-struct AggregateFuncTraits<OLAP_FIELD_AGGREGATION_NONE, field_type> {
-    static void aggregate(char* left, const char* right, Arena* arena) {}
+private:
+    void (*_init_fn)(char* dst, Arena* arena);
```

Comment:
may be use void* ?

Reply:
Change to void* will modify too many code. Because we put null ahead of data, we need +1 to get data pointer, which is a warning on void*. 
So I keep void* here, and will change it in later modification.